File: day14/Python/day14.py
from termcolor import colored
from pyfiglet import *
print(colored("Advent of Code - Day 14", "yellow").center(80, "-"))
print(colored(figlet_format("Extended Polymerization",font="small",justify="center"), 'green'))
print(colored("Output","yellow").center(80, "-"))

def ingestFile():
    f = open("../Input/day14.txt","r")
    lines = f.readlines()

    resultList = [x.strip() for x in lines]
    
    # Get the bingo number call list
    template = resultList[0].strip()

    cipher = {}

    for x in resultList[2:]:
        tmp = x.split(' -> ')
        cipher[tmp[0]] = tmp[0][0] + tmp[1] + tmp[0][1]

    return template, cipher

def genStringPairList(inString):
    tmpString = []
    for i in range(len(inString)):
        if i == (len(inString) - 2):
            tmpString.append(inString[i] + inString[i+1])
            break
        else:
            tmpString.append(inString[i] + inString[i+1])

    return tmpString


def part_1(template, cipher, steps):
    curString = template[:]

    for i in range(steps):
        newStringList = genStringPairList(curString)
        curString = ''
        for key, pair in enumerate(newStringList):
            newStringList[key] = cipher[pair]

        for key, value in enumerate(newStringList):
            if key == (len(newStringList) - 1):
                curString += value
            else:
                curString += value[:2]

    letterCount = dict()
    for i in curString:
        letterCount[i] = letterCount.get(i, 0) + 1


    maxElementCount = max(letterCount.values())
    minElementCount = min(letterCount.values())

    print(f'\nPuzzle 1: {maxElementCount - minElementCount}')

def part_2(template, cipher, steps):
    tmpList = genStringPairList(template)
    dPairs = {}

    result = {}
    for l in template:
        result[l] = result.get(l, 0) + 1

    for i in tmpList:
        dPairs[i] = dPairs.get(i,0) + 1

    for x in range(steps):
        tmpDict = {}
        for key, value in dPairs.items():
            tmpDict[cipher[key][:2]] = tmpDict.get(cipher[key][:2], 0) + value
            tmpDict[cipher[key][1:3]] = tmpDict.get(cipher[key][1:3], 0) + value
            result[cipher[key][1]] = result.get(cipher[key][1], 0) + value

        dPairs = tmpDict

    maxElementCount = max(result.values())
    minElementCount = min(result.values())

    print(f'Puzzle 2: {maxElementCount - minElementCount}')


template, cipher = ingestFile()

part_1(template, cipher, 10)

part_2(template, cipher, 40)
print()
print(colored("=".center(71, "="), "yellow"))

File: day14/Kotlin/day14.kt
import java.io.*
fun print_day_14() {
    val yellow = "\u001B[33m"
    val reset = "\u001b[0m"
    val green = "\u001B[32m"
    println(yellow + "-".repeat(25) + "Advent of Code - Day 14" + "-".repeat(25) + reset)
    println(green)
    val process = Runtime.getRuntime().exec("figlet Extended Polymerization -c -f small")
    val reader = BufferedReader(InputStreamReader(process.inputStream))
    reader.forEachLine { println(it) }
    println(reset)
    println(yellow + "-".repeat(33) + "Output" + "-".repeat(33) + reset + "\n")
}
fun main() {
    print_day_14()
    Day14().part_1()
    Day14().part_2()
    println("\n" + "\u001B[33m" + "=".repeat(72) + "\u001b[0m" + "\n")
}

class Day14 {
    val lines = File("../Input/day14.txt").readLines()
    val polyTemplate = lines.first()

    fun part_1() {
        val pairInsertions = lines.drop(2).map {
            val (pair, insertion) = it.split(" -> ")
            pair to insertion
        }

        var polyIteration = polyTemplate
        for (step in 1..10) {
            polyIteration = polyIteration.fold("") { str, curr ->
                val pairCheck = "" + (str.lastOrNull() ?: "") + curr
                val insert = pairInsertions.find { it.first == pairCheck } ?.second ?: ""

                str + insert + curr
            }
        }

        val elementCounts = polyIteration.groupBy { it }.map { it.key to it.value.size }.sortedBy { it.second }
        val ans = elementCounts.last().second - elementCounts.first().second
        println("Puzzle 1: $ans")
    }

    fun part_2() {
        val insertPattern = lines.drop(2).map { it.split(" -> ") }.associate { it[0] to it[1] }
        var pairFrequency = mutableMapOf<String, Long>()
        polyTemplate.windowed(2).forEach{ pairFrequency.put(it, pairFrequency.getOrDefault(it, 0) + 1) }
        for (step in 1..40) {
            val updatedFrequencies = mutableMapOf<String, Long>()
            pairFrequency.forEach {
                val key1 = it.key[0] + insertPattern[it.key]!!
                val key2 = insertPattern[it.key]!! + it.key[1]
                updatedFrequencies.put(key1, updatedFrequencies.getOrDefault(key1, 0) + it.value)
                updatedFrequencies.put(key2, updatedFrequencies.getOrDefault(key2, 0) + it.value)
            }
            pairFrequency = updatedFrequencies
        }
        val charFrequency = pairFrequency.toList().fold(mutableMapOf<Char, Long>()) { acc, pair ->
            acc.put(pair.first[0], acc.getOrDefault(pair.first[0], 0) + pair.second)
            acc
        }
        charFrequency.put(polyTemplate.last(), charFrequency.getOrDefault(polyTemplate.last(), 0) + 1)
        val sorted = charFrequency.values.sorted()
        val ans = sorted.last() - sorted.first()
        println("Puzzle 2: $ans")
    }
}